papis/commands/doctor.py

```diff
@@ -0,0 +1,374 @@
+    for example.
+    """
+    REGISTERED_CHECKS[name] = Check(name=name, operate=check_function)
+
+
+def registered_checks_names() -> List[str]:
+    return list(REGISTERED_CHECKS.keys())
+
+
+FILES_CHECK_NAME = "files"
+
+
+def files_check(doc: papis.document.Document) -> List[Error]:
+    """
+    It checks wether the files of a document actually exist in the
+    filesystem.
+    """
+    files = doc.get_files()
+    results = []  # type: List[Error]
+    folder = doc.get_main_folder()
+
+    def _fix(_file: str) -> Callable[[], None]:
+        """
+        Files fixer function, it will remove the bad file.
+        Notice that for now it only works if the file name is not of
+        the form 'subdirectory/file' but only 'file'.
+        """
+
+        def __fix() -> None:
+            _db = papis.database.get()
+            basename = os.path.basename(_file)
+            if basename in doc["files"]:
+                doc["files"].remove(basename)
+            doc.save()
+            _db.update(doc)
+
+        return __fix
+
+    for _f in files:
+        if not os.path.exists(_f):
+            results.append(Error(name=FILES_CHECK_NAME,
+                                 path=folder or "",
+                                 msg=("File '{}' declared but does not exist"
+                                      .format(_f)),
+                                 suggestion_cmd=("papis edit --doc-folder {}"
+                                                 .format(folder)),
+                                 fix_action=_fix(_f),
+                                 payload=_f,
+                                 doc=doc))
+    return results
+
+
+KEYS_EXIST_CHECK_NAME = "keys-exist"
+
+
+def keys_check(doc: papis.document.Document) -> List[Error]:
+    """
+    It checks wether the keys provided in the configuration
+    option ``doctor-keys-check`` exit in the document.
+    """
+    keys = papis.config.getlist("doctor-keys-exist-keys")
+    folder = doc.get_main_folder()
+    results = []  # type: List[Error]
+    for k in keys:
+        if k not in doc or len(doc[k]) == 0:
+            results.append(Error(name=KEYS_EXIST_CHECK_NAME,
+                                 path=folder or "",
+                                 msg=("Key '{}' does not exist"
+                                      .format(k)),
+                                 suggestion_cmd=("papis edit --doc-folder {}"
+                                                 .format(folder)),
+                                 fix_action=lambda: None,
+                                 payload=k,
+                                 doc=doc))
+    return results
+
+
+REFS_CHECK_NAME = "refs"
+
+
+def refs_check(doc: papis.document.Document) -> List[Error]:
+    """
+    It checks that a ref exists and if not it
+    tries to create one according to the ref-format configuration
+    if the user chooses to fix it.
+    """
+    folder = doc.get_main_folder()
+    bad_symbols = re.compile(r"[ ,{}\[\]@#`']")
```

Comment:
Is there some documentation for what characters are allowed (by bibtex?)? Didn't find anything from a short search..

Reply:
that is a good question, I also don't know for sure, but these are not kosher

Reply:
I remember trying to find out about those characters for my neovim plugin, and the best i could find is [this stackexchange thread](https://tex.stackexchange.com/questions/581901/what-is-the-safe-character-set-for-a-bibtex-label?noredirect=1&lq=1). It seems there isn't a clear set of characters, and what works depends on what software the `.bib` files are processed with.

Reply:
Yeah, that seems to be the best explanation. From a bit of digging
* `bibtexparser` (which we're using) seems to accept anything besides a space at https://github.com/sciunto-org/python-bibtexparser/blob/19e8dfe05490f2fb8ea1a9cc593090400eaea42a/bibtexparser/bibtexexpression.py#L160
* `btparse` (which is used by `biber` and `biblatex` and probably others) has an allowed list of characters at https://metacpan.org/dist/Text-BibTeX/view/btparse/doc/btparse.pod (under *name*)

But like that stackexchange post says, different languages may expand some non-alphanumeric characters and lead to interesting results when actually putting things into a `tex` file..

Reply:
Either way, the current choice is reasonable, so I'm perfectly fine with leaving this as is!